vyper/venom/analysis/available_expression.py

```diff
@@ -0,0 +1,272 @@
+from collections import deque
+from dataclasses import dataclass
+from functools import cached_property
+
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysesCache, IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.basicblock import (
+    BB_TERMINATORS,
+    IRBasicBlock,
+    IRInstruction,
+    IROperand,
+    IRVariable,
+)
+from vyper.venom.context import IRFunction
+from vyper.venom.effects import EMPTY, Effects
+
+_MAX_DEPTH = 5
+_MIN_DEPTH = 2
+
+UNINTERESTING_OPCODES = ["store", "param", "offset", "phi", "nop"]
+_IDEMPOTENT_INSTRUCTIONS = ["log", "call", "staticcall", "delegatecall", "invoke"]
```

Comment:
should be NONIDEMPOTENT_INSTRUCTIONS. also i think we should do it the other way. we have to explicitly mark an instruction as idempotent, otherwise it defaults to not-idempotent.